### ISSUE TITLE
.transaction([]) fails with Request Timeout Error on large dataset

### ISSUE THREAD

**Issue by durdenx** [prisma/prisma#3892]
## Bug description

I get this error in v2.8.1

```
_p [Error]: 
Invalid `prisma.injection.upsert()` invocation in
webpack-internal:///./src/pages/api/injection/index.ts:191:79


  Request Timeout Error
    at Jw.request (/Users/user/project/node_modules/@prisma/client/runtime/index.js:210:54)
    at processTicksAndRejections (internal/process/task_queues.js:97:5) {
  code: 'UND_ERR_REQUEST_TIMEOUT',
  meta: undefined
}
{
  "code": "UND_ERR_REQUEST_TIMEOUT"
}
```

if I use `prisma.$transation` on a large dataset (25k) and if i use `Promise.all` on the same dataset it works.

## How to reproduce

```ts
  const businesses= await prisma.business.findMany({
      where: { businessOptions: { some: { isSelected: true } } },
      // take: 10,
      include: {
        businessOptions: { where: { isSelected: true }, include: { directory: true } },
      },
    });

   const operations = prisma.map(({ ...data }) =>
      prisma.injection.upsert({
        where: { id: data.id },
        create: { ...data, business: { connect: { id: data.id } } },
        update: {
          ...data,
        },
      }),
    );

   await appClient.$transaction(operations);
```

## Expected behavior

Not throwing error when using `prisma.$transation` on a large dataset

## Environment & setup

- OS: Macos
- Database: PostgresSQL 12
- Node.js version: 12
- Prisma version: 2.8.1

**Comment by GlennChon** [comment]
I don't know if it's related but if I map an array with 325 items to upsert functions it'll work, if the array I map from has 326 then I get the request timeout error.

`Gf [Error]: 
Invalid `prisma.job.upsert()` invocation in
D:\Code\Javascript\api\src\bot.ts:51:15

  47    }
  48 })
  49 let cut = filtered.slice(0, 325)
  50 const upsertFuncs: any = cut.map((job,[object Object], any) =>
→ 51    prisma.job.upsert(


  Request Timeout Error
    at $w.request (D:\Code\Javascript\api\node_modules\@prisma\client\runtime\src\runtime\getPrismaClient.ts:1258:15)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async Promise.all (index 0) {
  code: 'UND_ERR_REQUEST_TIMEOUT',
  clientVersion: '2.9.0',
  meta: undefined
}
`
I was playing around with it in order to figure out what was causing this and I think it's the size of the data being passed to the upsert.  When I change the source of my array, the limit increases (possibly because the objects in the array contain less data, another possibility is that it has something to do the number being created and updated. Not entirely sure here).  The 2nd time I tried it, 334 array items worked but 335 failed.

**Comment by jshimko** [comment]
I have a job queue that runs an import of 25-30k records every day and I get this error 100% of the time now.  It actually fails when deleting the previous day's records before the insert (`deleteMany` call).  Also, it worked fine on Prisma 1.5.1, but we just upgraded to 1.9 (and fixed our breaking changes that blocked the last few updates) and now our nightly import job fails every time.

Do you still need a reproduction for this or is this a confirmed issue?  Happy to put something together, but I suspect any large insert/update/delete will trigger it because a bulk delete of 25k records does the same thing.  
 
Database: postgres:13-alpine
Node: 12.18.4
Prisma: 2.9.0

```
Request Timeout Error
  at $w.request (/Users/me/myapp/node_modules/@prisma/client/runtime/src/runtime/getPrismaClient.ts:1258:15)
  at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

**Comment by janpio** [comment]
A reproduction we could try would be wonderful.

**Comment by jshimko** [comment]
Ok @janpio, here's a reliable reproduction (at least for me)...

https://github.com/jshimko/prisma-issue-3892

That fails for me 100% of the time at around 25-30secs.  Same error...

```sh
Gf [Error]:
Invalid `prisma.user.create()` invocation in
/Users/me/prisma-issue-3892/script.ts:25:37

  21 }
  22
  23 // insert generated data
  24 await prisma.$transaction(
→ 25   users.map((data) => prisma.user.create(
  Request Timeout Error
    at $w.request (/Users/me/prisma-issue-3892/node_modules/@prisma/client/runtime/src/runtime/getPrismaClient.ts:1258:15)
    at processTicksAndRejections (internal/process/task_queues.js:97:5) {
  code: 'UND_ERR_REQUEST_TIMEOUT',
  clientVersion: '2.9.0',
  meta: undefined
}
```

Note that lowering the number to 10k records seems to work, so the script is confirmed valid other than the timeout issue.

Let me know if I can provide any more info.  Thanks!

**Comment by divyenduz** [comment]
@jshimko I am able to reproduce this, thanks for the reproduction.

**Comment by jshimko** [comment]
Hey guys.  Just curious for planning purposes on our side... is there currently any idea when this issue might be getting looked at (and worked on)?  Not being able to do large bulk inserts is actually a pretty significant blocker on our end that can't wait too much longer and our only temporary fix appears to be reverting back to 1.5.1 (pretty sure that was the last version that worked) and then undo our usage of several features that came in releases after that (which wouldn't be a trivial amount of work at this point).

I completely understand that you guys obviously have to favor your own business priorities.  I'm just trying to figure out if this is days, weeks, or months away.  Because if it's on the longer side, I'm going to have to revert a lot of updates on our side and go back to an older version of Prisma so we can get things working.  In the spirit of supporting open source projects that I depend on (and _really_ love working with!), I'm also happy to figure it out and submit a PR if that makes sense.  But I suspect that someone more familiar with the inner workings of Prisma would be a better/faster option.  

Let me know what you guys are thinking and I'll do whatever I have to on my side.  Thanks!

**Comment by Stevenje** [comment]
Getting this issue too when adding an array with 600 items in it.

**Comment by vicanera** [comment]
+1 for this.

**Comment by NikoMontana** [comment]
+1 - wondering why this issue is closed.

**Request Timout Error**

```
  let dbTransaction = [];

  for (let index = 0; index < articlesArray.length; index++) {

    const dbWrite = prisma.article.create({
      data: {
        artNr: article.artnr,
        imgSrc: article.bild,
        desc: _desc,
        shop: { connect: { name: shopName } },
      },
    });

    dbTransaction.push(dbWrite);
  }

  console.log('start transaction');
  const dbRes = await prisma
    .$transaction(dbTransaction)
    .then(() => console.log('transaction commited'))
    .catch((e) => console.log(e));
```

```
  code: 'UND_ERR_REQUEST_TIMEOUT',      
  clientVersion: '2.10.1',
  meta: undefined

TypeError: argument entity is required at etag 
```

using PSQL 12 & NextJS 10

**Comment by GlennChon** [comment]
@NikoMontana I think the most recent version has removed those errors.
You're using "2.10.1"

Upgrade to:
"@prisma/client": "^2.11.0",
"@prisma/cli": "^2.11.0",

**Comment by pantharshit00** [comment]
Yes, 2.11 removed timeouts from undici via https://github.com/prisma/prisma/pull/4156 so this is fixed in 2.11

**Comment by NikoMontana** [comment]
@GlennChon & @pantharshit00 thank you, updating to 2.11 solved it.
